_posts/2019-08-18-view-encapsulation-w-angularze.md

```diff
@@ -0,0 +1,426 @@
+```
+
+![ViewEncapsulation.None](/assets/img/posts/2019-xx-xx-view-encapsulation/view_encapsulation_none_result.jpg)
+
+Jak widzimy, wszystkie style paragrafów zostału dodane w sekcji `<head>`, co spowodowalo nadpisanie tego stylu ostatnim - `color: blue`. W efekcie wszystkie paragrafy mają ten sam kolor, również paragraf z komponentu **app-green**, który nie posiada żadnego stylu i powinien mieć domyślny kolor.
+
+### 2. ViewEncapsulation.Emulated (default)
+
+Domyślny tryb kapsułkowania w Angularze, w którym style są domknięte w komponencie.  
+W tym trybie style również znajdują się z sekcji `<head>`, ale posiadają dodatkowe atrybuty które wiążą je ze znacznikami HTML pochodzącymi z tego samego komponentu.  
+Dzięki temu na stronie możem istnieć kilka komponentów zawierających znacznik tego samego typu, ale z różnymi stylami.  
+**Uwaga!** W tym trybie style rodzica nie mają wpływu na elementy dziecka (ponieważ każdy element otrzymuje własny, unikalny atrybut).
+
+```typescript
+import {Component} from '@angular/core';
+
+@Component({
+    selector: 'app-red',
+    template: `
+        <p>Red paragraph!</p>
+    `,
+    styles: [`
+        p {
+            color: red;
+        }
+    `]
+})
+export class RedComponent {
+}
+```
+
+```typescript
+import {Component} from '@angular/core';
+
+@Component({
+    selector: 'app-green',
+    template: `
+        <p>Green paragraph!</p>
+    `,
+    styles: [`
+        p {
+            color: green;
+        }
+    `]
+})
+export class GreenComponent {
+}
+```
+
+```typescript
+import {Component} from '@angular/core';
+
+@Component({
+    selector: 'app-blue',
+    template: `
+        <p>Blue paragraph!</p>
+    `,
+    styles: [`
+        p {
+            color: blue;
+        }
+    `]
+})
+export class BlueComponent {
+}
+```
+
+```html
+<head>
+    <style>
+        p[_ngcontent-dql-c0] {
+            color: red;
+        }
+    </style>
+    <style>
+        p[_ngcontent-dql-c1] {
+            color: green;
+        }
+    </style>
+    <style>
+        p[_ngcontent-dql-c2] {
+            color: blue;
+        }
+    </style>
+</head>
+
+<body>
+    <app-root ng-version="8.2.2">
+        <app-red _nghost-dql-c0="">
+            <p _ngcontent-dql-c0="">Red paragraph!</p>
+        </app-red>
+        <app-green _nghost-dql-c1="">
+            <p _ngcontent-dql-c1="">Green paragraph!</p>
+        </app-green>
+        <app-blue _nghost-dql-c2="">
+            <p _ngcontent-dql-c2="">Blue paragraph!</p>
+        </app-blue>
+    </app-root>
+</body>
+```
+
+![ViewEncapsulation.Emulated](/assets/img/posts/2019-xx-xx-view-encapsulation/view_encapsulation_emulated_and_shadow_dom_result.jpg)
+
+Domyślny tryb pozwolił nam odseparować style między poszczególnymi komponentami, dzięki czemu uzyskaliśmy oczeiwany efekt - każdy paragraf ma swój kolor zdefiniowany w stylach komponentu.
+
+### 3. ViewEncapsulation.ShadowDom
+
+Kapsułkowanie oparte na **Shadow DOM** (wymaga wsparcia przeglądarki dla Shadow DOM).
```

Comment:
może warto coś napomknąć o tym które przeglądarki na razie to wspierają?

Reply:
tu jest fajna grafika, którą mozna by wrzucić: https://www.webcomponents.org/ (wykorzystywałem już przy custom elements: https://blog.consdata.tech/2019/03/27/custom-elements.html)

Reply:
![Zaznaczenie_092](https://user-images.githubusercontent.com/1863167/63354548-ebb27d80-c364-11e9-94bb-c066a1d5c6e2.png)
Nawet Ci wyciąłem graficzkę ;)

Wsparcie Shadow DOM przez główne przeglądarki
źródło: www.webcomponents.org - dostęp: 2019-08-20

Reply:
done